src/outputs/ascent.cpp

```diff
@@ -0,0 +1,199 @@
+//========================================================================================
+// Parthenon performance portable AMR framework
+// Copyright(C) 2023 The Parthenon collaboration
+// Licensed under the 3-clause BSD License, see LICENSE file for details
+//========================================================================================
+// (C) (or copyright) 2023. Triad National Security, LLC. All rights reserved.
+//
+// This program was produced under U.S. Government contract 89233218CNA000001 for Los
+// Alamos National Laboratory (LANL), which is operated by Triad National Security, LLC
+// for the U.S. Department of Energy/National Nuclear Security Administration. All rights
+// in the program are reserved by Triad National Security, LLC, and the U.S. Department
+// of Energy/National Nuclear Security Administration. The Government is granted for
+// itself and others acting on its behalf a nonexclusive, paid-up, irrevocable worldwide
+// license in this material to reproduce, prepare derivative works, distribute copies to
+// the public, perform publicly and display publicly, and to permit others to do so.
+//========================================================================================
+//! \file ascent.cpp
+//  \brief Ascent situ visualization and analysis interop
+
+#include <cstdio>
+#include <cstdlib>
+#include <iomanip>
+#include <iostream>
+#include <sstream>
+#include <stdexcept>
+#include <string>
+
+// Parthenon headers
+#include "coordinates/coordinates.hpp"
+#include "defs.hpp"
+#include "globals.hpp"
+#include "mesh/mesh.hpp"
+#include "outputs/outputs.hpp"
+
+// Ascent headers
+#ifdef PARTHENON_ENABLE_ASCENT
+#include "ascent.hpp"
+#include "conduit_blueprint.hpp"
+#include "conduit_relay_io.hpp"
+#include "conduit_relay_io_blueprint.hpp"
+#endif // ifdef PARTHENON_ENABLE_ASCENT
+
+namespace parthenon {
+
+//----------------------------------------------------------------------------------------
+//! \fn void AscentOutput:::WriteOutputFile(Mesh *pm)
+//  \brief  Expose mesh and all Cell variables for processing with Ascent
+void AscentOutput::WriteOutputFile(Mesh *pm, ParameterInput *pin, SimTime *tm,
+                                   const SignalHandler::OutputSignal signal) {
+#ifndef PARTHENON_ENABLE_ASCENT
+  if (Globals::my_rank == 0) {
+    PARTHENON_WARN("Ascent output requested by input file, but Ascent support not "
+                   "compiled in. Skipping this output type.");
+  }
+#else
+  // reference:
+  // https://llnl-conduit.readthedocs.io/en/latest/blueprint_mesh.html#complete-uniform-example
+
+  using conduit::Node;
+
+  // Ascent needs the MPI communicator we are using
+  ascent::Ascent a;
+  Node ascent_opts;
+  ascent_opts["mpi_comm"] = MPI_Comm_c2f(MPI_COMM_WORLD);
+  a.open(ascent_opts);
```

Comment:
This is probably not the main cause of the performance slowness, but we could move this to a higher-level scope and initialize at the start of a simulation, rather than re-creating the `ascent::Ascent object` and passing the MPI communicator each time. Since it's only used in this function, we could alternatively make this a static variable and initialize it only upon first use.

Reply:
Ah, great! Thanks for the input.
I was actually wondering about that but didn't find an info in their doc on whether it's fine to have a persistent object or if the model is "sim pause -> ascent start from scratch -> ascent data -> ascent done and cleanup -> sim resume".
Do you know if we could also reuse the nodes, i.e., caching them as long as the mesh structure didn't change?